### PR TITLE
sharding rocksdb first

### PR DESCRIPTION
**Pull request by mmucklo**

### CODE REVIEW

File: server/rocksdb_container.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "absl/strings/string_view.h"
+#include "rocksdb/db.h"
+
+namespace witnesskvs::server {
+
+class RocksDBContainer {
+  public:
+    // Creates a RocksDBContainer with num shards.
+    // Requirement: shards must be a power of 2.
```
Comment (vramesh7) [inline server/rocksdb_container.h:16]: Why ?

File: server/rocksdb_container.cc [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  CHECK_GT(f, 1);
+
+  // uint64_t for future expansion.
+  uint64_t shard_id = 0;
+  char x;
+  ;
+  size_t round = 1;
+  for (int i = 0; i < f - 1; i++) {
+    if (i % 8 == 0) {
+      x = key[key.length() - round];
```
Comment (rsingh16) [inline server/rocksdb_container.cc:65]: is round guranteed to be smaller than key length?

File: server/rocksdb_container.h (hunk 1 above)
Comment (vramesh7) [reply server/rocksdb_container.h:16]: Okay, nvm understood based on GetShardId()

File: server/rocksdb_container.cc [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    return 0;
+  }
+
+  int f = ffs(shards);
+  CHECK_GT(f, 1);
+
+  // uint64_t for future expansion.
+  uint64_t shard_id = 0;
+  char x;
+  ;
```
Comment (vramesh7) [inline server/rocksdb_container.cc:61]: stray ; ?

**Review by vramesh7 — APPROVED** [review]
Nice change

File: server/rocksdb_container.h (hunk 1 above)
Comment (mmucklo) [reply server/rocksdb_container.h:16]: We're doing bit-reversal of the string to find the shard id. Powers of two makes it easier when dealing with bits...

File: server/rocksdb_container.cc (hunk 2 above)
Comment (mmucklo) [reply server/rocksdb_container.cc:69]: good point, only matters for small keys and large # of shards (256 or greater), fixed.

TODO(mmucklo): write a large shard unit test.

File: server/rocksdb_container.h (hunk 1 above)
Comment (mmucklo) [reply server/rocksdb_container.h:16]: commented.

**Pull request merged**